### ISSUE TITLE
Export to zip gets stuck partway through: "Resolving mods"

### ISSUE THREAD

**Issue by allanonmage** [ryanccn/moddermore#301]
**Describe the bug**
I made a modpack yesterday, and when I tried to download it today to test it, I can't.  The pack contains a mix of Modrinth and Curseforge, so I'm using the export as zip functionality.

**To Reproduce**
1. Export as zip
2. Gets stuck.

On Waterfox, it gets stuck at 175 or 176.  On Vivaldi, it gets stuck at 325

**Expected behavior**
An output zip file, or an error message explaining what the problem is.

**Desktop (please complete the following information):**

- OS: Windows 10
- Browser Waterfox, Vivaldi

**Additional context**
Modpack: https://moddermore.net/list/2eX8EjJqfNKG

**Comment by allanonmage** [comment]
I saw a bunch of errors in F12 in Waterfox.  I'd post it, but it's too big for Github, so I put it on PasteBin: https://pastebin.com/a4xCavC5

**Comment by ryanccn** [comment]
https://github.com/modrinth/labrinth/issues/874

**Comment by allanonmage** [comment]
This is still happening.  I thought you or I had gotten a response from Modrinth that "you're doing it wrong", but I don't see that in your bug report above.  Is there another one somewhere?

**Comment by allanonmage** [comment]
Once they implement on 2951, do you need to do anything on Moddermore to take advantage of it?

**Comment by ryanccn** [comment]
@allanonmage

This issue is honestly quite hard to pin down. I've tried exporting your list on Firefox and Chromium; it seems to succeed more or less reliably on Chromium, while it often fails at an arbitrary number of mods on Firefox. Firefox apparently has a propensity to throw spurious CORS errors (thereby breaking the export process) on some requests when it is being rate limited, while Chromium does not. Moddermore has a rate limit backoff mechanism where it waits a few seconds before trying again, but currently this mechanism only activates when it receives proper response information from the browser, which Firefox does not seem too keen on providing. I can try to make the implementation more lenient towards browser-incurred errors, but for now I would suggest laying off Firefox if you are still using it.

With regards to https://github.com/modrinth/code/issues/1398#issuecomment-2267100034, I am not aware of any Modrinth API endpoint that allows batch queries for versions given project IDs and loader/game version information. So that is unfortunately not a solution.

**Comment by allanonmage** [comment]
Yeah, I do prefer Firefox, and use a flavor of it. I don't like Chrome, so I don't use it. I find it painfully ironic that Chrome is essentially the new IE 6 where websites only work correctly in it (even if this particular case is actually about a Firefox failing). I'll give it a shot on a Chrome-based browser though and try it again. Thanks for investigating it!

**Comment by ryanccn** [comment]
I've attempted a fix, see if it helps; I tried and it seems to export fine now for me.

**Comment by ryanccn** [comment]
That's good to hear! Closing this issue for now, but do let me know if you run into any more issues.
